### DISCUSSION TITLE
How do I log from cache

### DISCUSSION THREAD

**Discussion by dayadam** [App-vNext/Polly#1921] [Q&A] [answered]
```
        AsyncCachePolicy<TResult> GetCacheOnlyPolicy<TResult, TSerialized>(
            string cacheKey, 
            ICacheItemSerializer<TResult, TSerialized> serializer, 
            Func<TResult, bool> shouldCache, 
            IDistributedCache distributedCache) =>
            Policy.CacheAsync(
                distributedCache.AsAsyncCacheProvider(serializer),
                new ResultTtl<TResult>(result =>
                {
                    //Logger.LogInformation(true, "Result {result}.", result);
                    return shouldCache(result) ? new Ttl(CurrentPolicyOptionRecord.CacheTtl ?? default) : default;
                }),
                _ => cacheKey,
                (context, key) =>
                {
                    Logger.LogDebug(context.ToString());
                    Logger.LogDebug(true, "Cache hit {key}.", key);
                    //Logger.LogDebug(true, "Result being put into the cache: {result}.", result);
                },
                (context, key) =>
                {
                    Logger.LogDebug(context.ToString());
                    Logger.LogDebug(true, "Cache miss {key}.", key);
                    //Logger.LogDebug(true, "Result being put into the cache: {result}.", result);
                },
                (context, key) =>
                {
                    Logger.LogDebug(context.ToString());
                    Logger.LogDebug(true, "Cache put {key}.", key);
                    //Logger.LogDebug(true, "Result being put into the cache: {result}.", result); 
                },
                (context, key, exception) => Logger.LogWarning(true, exception, "Cache error getting {key}.", key),
                (context, key, exception) => Logger.LogWarning(true, exception, "Cache error putting {key}.", key));
```

How do I log what is being put into or gotten from the cache?

**Comment by peter-csala** [comment]
- [`onCachePut`](https://github.com/App-vNext/Polly/wiki/Cache#oncacheput)
> An optional `onCachePut` delegate allows specific code to be executed (for example for logging), after a value has been put to the cache.
- [`onCacheGet`](https://github.com/App-vNext/Polly/wiki/Cache#oncacheget)
> An optional `onCacheGet` delegate allows specific code to be executed (for example for logging), when a value is retrieved from cache.

[Here](https://github.com/App-vNext/Polly/blob/51044c56ff0d76c470df78a5bc03e20629e250ca/src/Polly/Caching/CacheEngine.cs) you can see when these above two delegates are being called.

___

```cs
Policy.Cache(
      cacheProvider: /* the chosen cache provider */,
      ttl: /* time to live timespan */,
      onCacheGet: (_, key) => Logger.Info("Hit {key}", key),
      onCacheMiss: (_, __) => {},
      onCachePut: (_, key) => Logger.Info("Put {key}", key),
      onCacheGetError: null,
      onCachePutError: null
);
```

**Reply by dayadam** [reply]
That just logs the key, not the value

**Reply by peter-csala** [reply]
`cacheProvider.TryGet(key)` would retrieve the data from the cache. Either use closure to capture the provider or pass it through the context to the delegates.

**Reply by dayadam** [reply]
Thank you. Let's say I do that on cache put, would that be what is in the cache before or after the put?

**Reply by peter-csala** [reply, marked as answer]
@dayadam After the put. See line 56 and 57: https://github.com/App-vNext/Polly/blob/51044c56ff0d76c470df78a5bc03e20629e250ca/src/Polly/Caching/CacheEngine.cs#L56
https://github.com/App-vNext/Polly/blob/51044c56ff0d76c470df78a5bc03e20629e250ca/src/Polly/Caching/CacheEngine.cs#L57

**Reply by dayadam** [reply]
I'm having a really hard time getting the content of the value. When I debug and the type is TResult, I can see headers and there is content from the response availble. I can't access content from type TResult as it's not define and I can't seem to implement a custom class (is this what I need to do somehow?). But once I convert to HTTPResponseMessage the content disappears. How can I access the content of type TResult or convert it without data loss?

**Reply by peter-csala** [reply]
Sorry, could you please provide a bit more context? Or could you please share the related code fragment?

**Reply by peter-csala** [reply]
@dayadam 

Reading your question after a coffee I think I know what do you want. With C# 9's enhanced pattern matching you can do something like this:

```cs
var response = new HttpResponseMessage() { Content = new StringContent("Hello") };

await ProcessGenericResult(response);

async Task ProcessGenericResult<TResult>(TResult response)
{
    if(response is HttpResponseMessage { Content: var content })
    {
        Console.WriteLine(await content.ReadAsStringAsync());
    }
}
```
So, if the `response` is an instance of `HttpResponseMessage` then capture its `Content` property into a variable called `content`. Inside the `if` statement you can easily read content. 

I hope it helped.
